library/sha256.c

```diff
@@ -22,8 +22,31 @@
  *  http://csrc.nist.gov/publications/fips/fips180-2/fips180-2.pdf
  */
 
+#if defined(__aarch64__) && !defined(__ARM_FEATURE_CRYPTO) && \
+    defined(__clang__) &&  __clang_major__ < 18 && __clang_major__ > 3
+/* TODO: Re-consider above after https://reviews.llvm.org/D131064 merged.
+ *
+ * The intrinsic declaration are guarded with ACLE predefined macros in clang,
+ * and those macros are only enabled with command line. Define the macros can
+ * enable those declaration and avoid compile error on it.
+ */
+#define __ARM_FEATURE_CRYPTO 1
+#pragma clang attribute push (__attribute__((target("crypto"))), apply_to=function)
```

Comment:
Unfortunately, I think there's a regression here. We're now forcing `target=crypto` in all Aarch64 builds with Clang, for any `static inline` function defined by a header. In builds with A64 disabled or if-present, these functions must not use target-specific instructions. Granted, it's unlikely that a function defined in a header would end up using target-specific instructions, but it makes me uneasy. Note that the target-specific instructions might not be limited to crypto-specific instructions: for example the library might be built without the use of Neon instructions (they're mandatory in the architecture IIRC, but they can be disabled for the sake of OSes that don't save the Neon registers).

Similarly below, for all the code in this file until the second `#pragma clang attribute pop` line, including functions like `mbedtls_sha256_starts`, we're allowing the compiler (Clang or GCC) to use target-specific instructions. There we only do it when some A64 acceleration is requested, so it's a defect in the `IF_PRESENT` case (the software implementation may end up using target-specific instructions), but that's not a regression.

I really think we should put the A64 code in its own file. It would have several advantages: remove the complexity of pragma popping; make it more obvious which code can use which instructions; allow users with other compilers to tune options in their build scripts.

Reply:
> Unfortunately, I think there's a regression here. We're now forcing `target=crypto` in all Aarch64 builds with Clang, for any `static inline` function defined by a header. In builds with A64 disabled or if-present, these functions must not use target-specific instructions. Granted, it's unlikely that a function defined in a header would end up using target-specific instructions, but it makes me uneasy. Note that the target-specific instructions might not be limited to crypto-specific instructions: for example the library might be built without the use of Neon instructions (they're mandatory in the architecture IIRC, but they can be disabled for the sake of OSes that don't save the Neon registers).

Yes, and I re-write this part.
> 
> Similarly below, for all the code in this file until the second `#pragma clang attribute pop` line, including functions like `mbedtls_sha256_starts`, we're allowing the compiler (Clang or GCC) to use target-specific instructions. There we only do it when some A64 acceleration is requested, so it's a defect in the `IF_PRESENT` case (the software implementation may end up using target-specific instructions), but that's not a regression.
> 
> I really think we should put the A64 code in its own file. It would have several advantages: remove the complexity of pragma popping; make it more obvious which code can use which instructions; allow users with other compilers to tune options in their build scripts.

Agree, but I do not think that's scope of this PR.  

I have not understand the structure of `sha{512,256}.c`. I prefer postpone this issue.